### PR TITLE
UNA - Implementing split greenspace

### PR DESCRIPTION
**Pull request by phargogh**
## Description

This PR implements the optional split greenspace feature of UNA, which is mostly a slight refactor of the 2SFCA procedure described in the [UNA design doc](https://docs.google.com/document/d/1lqx75-S605lLxingVjd_yVSGt1tssEFp/edit).

There are a couple of unresolved questions about this feature that I'm waiting to hear back from the science team about, specifically:

1. The design doc describes that `Ai,r` (greenspace supply per greenspace class) is output, but it is not clear if it's important for one such output to be produced for each greenspace classification or if these can be aggregated by a shared search radius as they are now, and are in the base model.
2. The way in which the user defines their various radii for split greenspace doesn't feel great to me in that there's some duplication of the input between a table and a separate UI element.  This may change in the future, and I'd like to ask the science team what they think about it.

This PR also includes all the workbench changes, so sorry for the big diff!  If this is impractical, I'll be happy to do a separate PR that's just `main` into `feature/urban-nature-access`.

## Checklist
~~- [ ] Updated HISTORY.rst (if these changes are user-facing)~~
~~- [ ] Updated the user's guide (if needed)~~
- [X] Tested the affected models' UIs (if relevant)

### CODE REVIEW

**Comment by emlys** [comment]
Hey @phargogh, it looks like there's a whole bunch of other changes in the diff not related to the purpose of this PR. Maybe need to separately merge `main` into `feature/urban-nature-access`?

**Comment by phargogh** [comment]
Thanks for taking a look @emlys !  I've created https://github.com/natcap/invest/pull/953 for merging `main` into `feature/urban-nature-access`.  Once that's in, this review should become a lot easier.

**Comment by emlys** [comment]
Hey @phargogh I'm still seeing 225 changed files in the diff even though #953 was merged. Do you know if there's a way to refresh the diff?

**Comment by emlys** [comment]
Awesome thank you!

**Comment by emlys** [comment]
> The way in which the user defines their various radii for split greenspace doesn't feel great to me in that there's some duplication of the input between a table and a separate UI element.

I agree. If someone wants to use the same radius for every LULC, it's easy enough to enter the same value for each row in their table. The separate radius input seems unnecessary.

File: src/natcap/invest/urban_nature_access.py
Comment (emlys) [inline src/natcap/invest/urban_nature_access.py:288]: I like how you're using all the args as `kwargs` even though they don't have to be. It's nice to see the arg variable name alongside its value.

File: src/natcap/invest/urban_nature_access.py
Comment (emlys) [inline src/natcap/invest/urban_nature_access.py:296]: Does setting the `priority=10` still have any effect if we `join` the task first?

File: src/natcap/invest/urban_nature_access.py
Comment (emlys) [inline src/natcap/invest/urban_nature_access.py:320]: I'm not sure what this is saying, maybe missing a word?

File: src/natcap/invest/urban_nature_access.py
Comment (emlys) [inline src/natcap/invest/urban_nature_access.py:520]: There are raster sum functions implemented in other models that you could import and reuse here, but I don't know if they're just as good. Given that these type of functions are duplicated a lot in invest, adding this one isn't a big deal. https://github.com/natcap/pygeoprocessing/issues/235

File: src/natcap/invest/urban_nature_access.py
Comment (emlys) [inline src/natcap/invest/urban_nature_access.py:560]: This is a little unclear to me. Maybe another line saying something like "Greenspace pixels' value will be the pixel area, and non-greenspace pixels' value will be 0." (If I'm understanding that correctly)

File: src/natcap/invest/urban_nature_access.py
Comment (emlys) [inline src/natcap/invest/urban_nature_access.py:636]: `sep=None` and `engine='python'` are the defaults, but feel free to leave them if you think it helps to be explicit here.

File: src/natcap/invest/urban_nature_access.py
Comment (emlys) [inline src/natcap/invest/urban_nature_access.py:871]: The default `isclose` tolerances are rtol=1e-05, atol=1e-08. And I think rtol would actually be 0 because the compared value is 0. I wonder if you might want an even higher atol since in the context of this model, 0.00000008 people is probably negligible. Maybe even as much as 0.5 since that would already double the "effective greenspace" (as I'm understanding this ratio).

File: tests/test_urban_nature_access.py
Comment (emlys) [inline tests/test_urban_nature_access.py:44]: Neat!

File: tests/test_urban_nature_access.py
Comment (emlys) [inline tests/test_urban_nature_access.py:412]: Should this go in a `tearDown`?

**Review by emlys — CHANGES_REQUESTED** [review]
@phargogh here's my first round of comments! I know you said you'd have a few more changes to make so I'll look over again when those are ready.

File: tests/test_urban_nature_access.py
Comment (dcdenu4) [inline tests/test_urban_nature_access.py:44]: `random.choices` - cool!

File: tests/test_urban_nature_access.py
Comment (dcdenu4) [inline tests/test_urban_nature_access.py:427]: That's interesting. I'm not plugged in to the details here, but I wonder if there needs to be an explicit enable/disable switch? With the way I think you have it set up, if the table includes the search radius column with different radii, then that is indicating a `split greenspace`. That seems right to me, with a LOGGER INFO indication of which mode is being used. But yea, it'd have to be articulated well in the Uses Guide.

**Review by dcdenu4 — CHANGES_REQUESTED** [review]
Hey @phargogh , this mostly looks good to me! I just had a few questions and comments. Sorry for the delay on this.

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:74]: I addressed this particular bit of whitespace in a larger refactor in b1542ce17

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:137]: I took a first stab at this in f611e112b.  I think the choices are pretty clear, though the user's guide should (and the design doc does) include the math for each function.

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:139]: Yep!  Added in f611e112b

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:76]: In talking with Roy and Hongxiao (and also with your feedback), this column is now required.

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:122]: This input is now removed.

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:239]: Yeah, I thought a bit about this and I think I'd prefer to keep them separate, just with an assertion to guard against programmer error.  I feel like keeping them separate ensures a separation of concerns: ARGS_SPEC info with the ARGS_SPEC, the mapping of kernel functions here in `execute`.

That being said, I did use a few more constants in 2ddbbe6b3 to represent the kernel keys (which cuts out one type of duplication).  The assertion is still useful here though, so I've left it in.

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:249]: Yep, sure does!  I've removed this check in b2795a8830157bbbd2387fe80cc3d5f56c4179af with the assumption that validation will handle it.  All cases are covered in ARGS_SPEC, so this shouldn't be hard to cover.

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:288]: Thanks!  I've found it really helpful and clear when writing this out and it helps cut down on issues where I accidentally pass in the wrong thing.

The next nice-to-have would be to have my editor populate the kwargs ... not there yet.

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:296]: In this case, the issue is addressed by no longer needing to preprocess the table since the `search_radius_m` column is now required.

In the general case, though, yes, setting `priority=10` could still have an effect when the task it `join`ed.  Imagine a case where `n_workers > 0`, which means that the order of task execution is not guaranteed.  It's possible for us to get into a case where all of the other tasks that were added other than this one task to execute before `preprocess_attr_table_task`.  Increasing the priority means that this task will take priority if another lower-priority task also has satisfied its dependencies.

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:320]: Ha, this really isn't readable at all, is it?  Greatly simplified in  cfa5bbdd9

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:520]: Yep, I was previously using NDR's implementation and then switched (I don't remember why).  I just switched back, which seems fine.

But yes, I'd really rather see this addressed in a shared utility.

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:560]: Added a description along these lines in 0dacced0e

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:636]: I think this function has been removed because this function is no longer needed.

File: tests/test_urban_nature_access.py
Comment (phargogh) [reply tests/test_urban_nature_access.py:412]: Well, really the whole test needed a refactor because the core model now also requires the search radius column, so the line is now removed!  b5a46055a

File: src/natcap/invest/ui/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/ui/urban_nature_access.py:46]: Yep, the science team agrees!  And I've removed this UI input in 1dbaa6d68

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:293]: When it's useful, it's useful!

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:296]: Yeah, @emlys asked the same question, so I hope I offered a good-enough explanation above: https://github.com/natcap/invest/pull/944#discussion_r864411498

But also, we no longer need this join!  So it (and the priority) have been removed.

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:819]: Sure should!  036520324

File: tests/test_urban_nature_access.py
Comment (phargogh) [reply tests/test_urban_nature_access.py:427]: Yeah, there might not need to be any more, if we assume that the presence of multiple search radii triggers the split greenspace mode!

I'll need to work on this a little more and I'll get back to you soon.

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:803]: Yeah, that's a really interesting point.  Let me think about this for a bit and I'll get back to you soon.

**Comment by phargogh** [comment]
Thanks a ton for your feedback @emlys and @dcdenu4 !  I've responded to most of your comments, but I've got a little more thinking to do on two of your comments:

- [ ] https://github.com/natcap/invest/pull/944#discussion_r859222795
- [ ] https://github.com/natcap/invest/pull/944#discussion_r862898760

I'll get back to you soon and re-request your review.

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:803]: Yep, I agree, I think a higher atol would be a really good idea.

The main reason is that the population raster coming in to this function is coming out of a non-normalized convolution, and so even within float64 space it wouldn't be that hard for the accumulated numerical error to exceed 1e-8.

In practice, I suspect this case will mostly not be triggered, given this is an urban model and that a single low-population pixel would need to be on the very fringes of a search kernel to get such a low number.  I tried running this on the Paris sample data with atols ranging from 1e-8, 1e-7 ... 1e-1, 1 with no detectable differences between the outputs, which isn't surprising for an urban area, even one surrounded by farmland and open space.

Yep, so you're right that if we had a super low population, we'd end up with a higher ratio since there'd be double (in the case of 0.5 people) the greenspace per person.

So in the interest of caution, how about we bump this up to 0.1?

Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:803]: In case you agree, I've done this and added a comment in 5275ee0bd

File: tests/test_urban_nature_access.py
Comment (phargogh) [reply tests/test_urban_nature_access.py:351]: OK! I've implemented this in https://github.com/phargogh/invest/commit/52fbbd565b776ba91b79eab0bf27de957f1d2803.  The behavior now is:

* The `search_radius_m` column is always required in the attributes table
* If the search radii are all the same, we treat that as "unified greenspace" and all greenspace lucodes will be processed together.  An INFO message is logged describing what's up.
* If there are multiple detected search radii, we use "split greenspace" mode and each individual LULC code is processed seprately (per request of the science team).  an INFO message is logged describing what's up.

As a part of this I also factored the kernel creation and population convolution out of the 2SFCA loop since there isn't necessarily a 1:1 mapping of search radii to greenspace lucodes.

**Comment by phargogh** [comment]
OK!  Well one thing led to another since your last kind review, so I'm sorry about the delay.  This is now ready for a re-review, I think, and thanks in advance!

**Comment by phargogh** [comment]
So the latest changes to `main` are also in this PR again.  I've opened up #1012 to help make this review easier to read.  Feel free to review that one first and once it's merged I'll update this PR.

**Comment by phargogh** [comment]
@emlys @dcdenu4 I just pushed a change in 807eb9bc2 that I think might help to clarify the radius/lucode groupings question, and it includes some more comments as requested.  Let me know if there's anything else you'd like to talk through!

File: src/natcap/invest/urban_nature_access.py
Comment (emlys) [inline src/natcap/invest/urban_nature_access.py:833]: I think this and the other `0.0`s can just be `0` since python 3

File: src/natcap/invest/urban_nature_access.py
Comment (emlys) [inline src/natcap/invest/urban_nature_access.py:849]: As we talked about in person, it might make more sense to set the ratio to the greenspace area for all population values in the range 0-1

Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:849]: Patching this in 3f68df5fd and a doc forthcoming!

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:833]: Yep, they sure don't: f8087c3ae

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:259]: I agree, this optional behavior just came back to bite me right now, where I was using the correct key names everywhere _except_ in `execute`, which led to incorrect regression values having been committed to the test.  I've fixed this and there's really no reason why I can't just require the kernel name too, so I am now.  See rev b2a1f7a59

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:77]: Patched in 03c91cd54

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py]: Here's a link to the doc for future us (stanford auth needed): https://docs.google.com/document/d/1CdzGT9YyeedtjaXQ17ZBGkg2KrsD8HTPUZ2sS7GZ3Uo/edit?usp=sharing

**Comment by dcdenu4** [comment]
I've been slow to respond and haven't allowed #1012 to be merged yet! Sorry! Once that is done this should be quick to merge in. (mostly note to self).

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py]: We heard back from the urban team and here's Lisa's response:

> James and Emily, I just chatted with Roy Remme (who had lost access to Slack, but is back now), and Anne and Eric, too, and we all agree your solution looks like a great one.And we really appreciate that you noticed the funky discontinuity implied by the original model!

So!  The approved solution is included in this PR already.

**Comment by dcdenu4** [comment]
Hey @phargogh, just a heads up that #1012 has been merged producing a few conflicts in this PR.

**Comment by phargogh** [comment]
OK!  Conflicts have been resolved and these branches are now much more up-to-date with the latest `main`.  Re-requesting review.

File: src/natcap/invest/urban_nature_access.py
Comment (dcdenu4) [inline src/natcap/invest/urban_nature_access.py:357]:
```suggestion
        # the cost of more convolutions and more runtime (and disk space).
```

Comment (dcdenu4) [reply src/natcap/invest/urban_nature_access.py:357]: I suggested a removal of that last bit because I'm not sure it's helpful. Let me know what you think!

File: src/natcap/invest/urban_nature_access.py
Comment (dcdenu4) [inline src/natcap/invest/urban_nature_access.py:361]: `groupby`, nice

**Review by dcdenu4 — CHANGES_REQUESTED** [review]
Thanks @phargogh , I just had one very minor suggestion on a comment. Otherwise this looks good to me!

File: src/natcap/invest/urban_nature_access.py
Comment (phargogh) [reply src/natcap/invest/urban_nature_access.py:360]: Yeah, that `groupby` is really kind of interesting ... it's yielding a generator of dataframes, which is _not_ immediately clear.

File: src/natcap/invest/__init__.py
Comment (emlys) [inline src/natcap/invest/__init__.py:35]: Possibly something amiss with the merge? This looks correct but I'm surprised this file is in the diff

**Review by emlys — APPROVED** [review]

**Review by dcdenu4 — APPROVED** [review]
Thanks for all of this @phargogh! I'm going to approve but let you do the final merge since the binary tests are failing (seemingly for different reasons), and Emily had a comment about something `gettext` related that raised a flag.

**Comment by phargogh** [comment]
So there are 2 issues for sure in the binary builds, one with the UG not building (pending https://github.com/natcap/invest.users-guide/pull/69, https://github.com/natcap/invest.users-guide/pull/68) and one with a bug in my workbench ui config for UNA.  I've pushed a fix for the workbench ui config and am waiting on tests now.

**Comment by phargogh** [comment]
OK!  We have 2 approvals and I think the remaining issue is related to the forums tag and will need to be figured out in due time.  We have 2 approvals now and these test failures appear minor so I'll go ahead and merge.

**Pull request merged**
